### PR TITLE
[WIP] Make '%A' structured formatting be the default for F# plaintext 

### PR DESCRIPTION
**Pull request by dsyme**
This extends over #689 to make '%A' formatting by the standard for plaintext rendering for F#

Addresses much of #642 

It's work in progress

### CODE REVIEW

File: src/Microsoft.DotNet.Interactive.FSharp.Tests/Microsoft.DotNet.Interactive.FSharp.Tests.fsproj [hunk 1]
```diff
@@ -14,6 +14,8 @@
     <Compile Include="ApiTests.fs" />
   </ItemGroup>
 
+  <ItemGroup />
```
Comment (jonsequitur) [inline src/Microsoft.DotNet.Interactive.FSharp.Tests/Microsoft.DotNet.Interactive.FSharp.Tests.fsproj:17]: Is this a VS artifact?

File: src/Microsoft.DotNet.Interactive.FSharp/FSharpKernelExtensions.fs [hunk 2]
```diff
@@ -17,35 +17,76 @@ open XPlot.Plotly
...
-        let code = @"
-open System
-open System.Text
-open System.Threading.Tasks
-open System.Linq"
+        // F# has its own views on what namespaces are open by default in its scripting model
+        kernel
+
+    [<Extension>]
+    static member UseExtraNamespacesForTesting(kernel: FSharpKernelBase) =
```
Comment (jonsequitur) [inline src/Microsoft.DotNet.Interactive.FSharp/FSharpKernelExtensions.fs:71]: This looks like it should be in the test project.

**Comment by dsyme** [comment]
@jonsequitur asked this

> I haven't noticed in your PRs any explicit separation of F# from C# plain text formatting. Is it something you've run into yet? The tricky piece here is that since they're in process together and currently formatters are not language-specific, which is by design. But I think we need to decide when and where it's appropriate to differ.

Yes, this is actively on my mind, and I'm trying to make as many improvements as possible which are language-independent. However F# box-layout plaintext formatting is a good mechanism we need to integrate, and there is no equivalent for C# as yet.

We could possible look at incorporating an adapted version of that mechanism into .NET Interactive for C# as well.  But that's further down the line.  

I think the order is

1. improve the common basis (what I'm doing now)

2. incorporate F#-specific box layout plaintext formatting (basically https://github.com/dotnet/interactive/pull/705)

3. make some improvements to  F#-specific box layout plaintext formatting in dotnet/fsharp

4. think about whether to duplicate/generalise the mechanism or something similar for C#

**Comment by dsyme** [comment]
So if we need F# plaintext formatting to be different (e.g. "1L" not "1"), then I guess that means we're going to have to start scoping the formatting settings to a particular kernel.

Currently everything is, well, very global....

In general how do we feel about this kind of thing?  e.g. a polyglot programmer can today register a formatter in C# and have it implicitly picked up by F#?  It seems likely better to avoid building in "global leakage" across kernels simply because C# and F# happen to be implemented using the same DLLs...

**Review by caroletchepare — APPROVED** [review]

**Comment by dsyme** [comment]
I'm going to close this out for now

When I looked at this I didn't appreciate that the defaults weren't specific to each language.  I may salvage parts of this when I next take notebooks for a long spin from the F# perspective

**Comment by dsyme** [comment]
I'm going to close this out for now

When I looked at this I didn't appreciate that the defaults weren't specific to each language.  I may salvage parts of this when I next take notebooks for a long spin from the F# perspective

**Pull request closed without merging**